system/src/commonMain/kotlin/com/splendo/kaluga/system/network/BaseNetworkManager.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ Copyright 2020 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+abstract class BaseNetworkManager(private val networkStateRepo: NetworkStateRepo) : CoroutineScope by networkStateRepo {
+
+    internal abstract suspend fun startMonitoringNetwork()
+    internal abstract suspend fun stopMonitoringNetwork()
+    internal abstract suspend fun isNetworkEnabled(): Boolean
+
+    internal fun handleNetworkStateChanged(network: Network) {
+        launch {
+            networkStateRepo.takeAndChangeState { state: NetworkState ->
+                when (state) {
+                    is NetworkState.Available -> {
+                        if (isNetworkEnabled()) {
+                            when (network) {
+                                is Network.Cellular -> state.availableWithCellular
+                                is Network.Wifi -> state.availableWithWifi
+                                else -> state.unavailable
+                            }
+                        } else  {
+                            state.unavailable
+                        }
+                    }
+                    is NetworkState.Unavailable -> {
+                        if (isNetworkEnabled()) {
+                            when (network) {
+                                is Network.Cellular -> state.availableWithCellular
+                                is Network.Wifi -> state.availableWithWifi
+                                else -> state.unavailable
+                            }
+                        } else  {
+                            state.remain()
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+expect class NetworkManager(networkStateRepo: NetworkStateRepo, context: Any? = null) : BaseNetworkManager
```

Comment:
Use a Builder to pass platform context. You can also use `ApplicationHolder`, probably location is the best example to look at.

Reply:
Is it fine if instead of creating a Builder I add a lazy property in android/NetworkManager that holds `ApplicationHolder.applicationContext`?

Reply:
No, `Context` should be a constructor parameter of the builder, but it can be filled by default by `ApplicationHolder.applicationContext`. I think location does the same again.

Reply:
4dad1f9 795f7a7 d521ff8 a42203b